crates/matrix-sdk/src/sliding_sync/mod.rs

```diff
@@ -483,21 +485,43 @@ impl SlidingSync {
             if self.inner.sticky.read().unwrap().data().extensions.e2ee.enabled == Some(true) {
                 debug!("Sliding Sync is sending the request along with outgoing E2EE requests");
 
-                let (e2ee_uploads, response) =
-                    futures_util::future::join(self.inner.client.send_outgoing_requests(), request)
-                        .await;
+                // If the sliding sync request fails faster than the e2ee requests get their
+                // responses, we want to fail as early as possible, so as to invalidate the
+                // sliding sync session as early as possible.
+                //
+                // Also, we don't want an interruption in the sliding sync request to abort
+                // processing of the e2ee response.
+                //
+                // For those reasons, start the e2ee requests in a background task. If this is
+                // aborted while everything is running, the same data might be sent later, which
+                // is fine.
+
+                let client = self.inner.client.clone();
+                let e2ee_uploads = spawn(async move {
+                    if let Err(error) = client.send_outgoing_requests().await {
+                        error!(?error, "Error while sending outoging E2EE requests");
+                    }
+                })
+                // Ensure that the task is not running in detached mode. It is aborted when it's
+                // dropped.
+                .abort_on_drop();
 
-                if let Err(error) = e2ee_uploads {
-                    error!(?error, "Error while sending outgoing E2EE requests");
-                }
+                // Wait on the sliding sync request success or failure early.
+                let response = request.await?;
+
+                // At this point, if `request` has been resolved successfully, we wait on
+                // `e2ee_uploads`. It did run concurrently, so it should not be blocking for too
+                // long. Otherwise —if `request` has failed— `e2ee_uploads` has
+                // been dropped, so aborted.
+                e2ee_uploads.await.map_err(|_| Error::JoinError("e2ee_uploads".to_string()))?;
```

Comment:
Please `to_owned()` instead of `to_string()`. But why throw away the `JoinError`?

Reply:
Damn, I keep doing this error. Memory muscle is hard. Thanks!

Reply:
I can't keep `JoinError` because `Error` implements `PartialEq` (for testing purposes), and `JoinError` doesn't implement `PartialEq`…

Let's implement `PartialEq` manually I guess…

Reply:
What, we implement `PartialEq` for errors? Let's not do that! Let's use `assert_matches!` in tests instead.

Reply:
Done :-).